# Basic Image Standards

In this notebooks we are going to explore some of the different standards that are used to store photographs and similar images. Specifically, we will explore the following standards

* [TGA](https://en.wikipedia.org/wiki/Truevision_TGA)
* [PNG](https://www.w3.org/TR/2003/REC-PNG-20031110/)
* [TIFF](https://www.adobe.io/open/standards/TIFF.html)
* [JPEG](https://jpeg.org/jpeg/)

In [2]:
from dminteract.modules.m4c import *


We are going to start with a sample photograph of my daughter (used with her permission!).

<img src="./data/daughter.jpg" alt="daughter with basketball" width="128" height="192">

Before we proceed, I want you to brainstorm about what be the minimum metadata you would need to describe this image.

### Type your thoughts here in a bulletted list

* Here is the 1st element of a list
* Here is the 2nd element of a list
* size
* contents (e.g., person, dog, building)
* color/b&w
* 

### Once you have typed your thoughts about the metadata...

Let's look at the metadata that is stored for our simplest ima

In [3]:
print(view_img_metadata("./data/daughter.tga"))


Metadata:
- Image width: 1998 pixels
- Image height: 3024 pixels
- Bits/pixel: 24
- Compression: True-color uncompressed
- MIME type: image/targa
- Endianness: Little endian


### Are there any elements that surprised you?

* Type your first answer here
* Type your second answer here
* ... and so on
* MIME type
* Endianness

## Here are some elements that stood out to me

* [Endianness](https://en.wikipedia.org/wiki/Endianness) is a detailed concept that I'm guessing many of you haven't been exposed to before.
* [MIME type](https://en.wikipedia.org/wiki/Media_type) is information used in transmitting binary data across the internet
* 24 bits per pixel: this is 8 bits per red, green, and blue channel that are used to represent color
* compression: this file format does not used data compression

### Now let's look at a the TIFF representation

In [4]:
print(view_img_metadata("./data/daughter.tiff"))

Metadata:
- Image width: 1998 pixels
- Image height: 3024 pixels
- Bits/pixel: 24
- Image DPI width: 72 DPI
- Image DPI height: 72 DPI
- MIME type: image/tiff
- Endianness: Little endian


### Any notable differences between TIFF and TGA?

* DPI width and height

### This is what I noticed

* There is now a DPI ("dots per inch") concept. TIFF was developed by Adobe which was focused on printing graphics.

## Now let's look at PNG

In [5]:
print(view_img_metadata("./data/daughter.png"))


Metadata:
- Image width: 1998 pixels
- Image height: 3024 pixels
- Bits/pixel: 24
- Pixel format: RGB
- Compression rate: 3.8x
- Image DPI width: 2834 DPI
- Image DPI height: 2834 DPI
- Compression: deflate
- Comment: Raw profile type iptc=
iptc
      35
1c015a00031b25471c0200000200021c0205000f696d675f3037303930355f30333634
- Comment: date:create=2020-03-25T23:34:30+00:00
- Comment: date:modify=2017-02-06T03:47:30+00:00
- Comment: exif:DateTime=2008:12:23 12:50:21
- Comment: exif:DateTimeDigitized=2005:07:08 19:51:47
- Comment: exif:DateTimeOriginal=2005:07:08 19:51:47
- Comment: exif:ExifOffset=200
- Comment: exif:ExifVersion=48, 50, 50, 48
- Comment: exif:ExposureBiasValue=0/1
- Comment: exif:ExposureProgram=3
- Comment: exif:ExposureTime=1/1250
- Comment: exif:Flash=0
- Comment: exif:FNumber=9/5
- Comment: exif:FocalLength=85/1
- Comment: exif:LightSource=9
- Comment: exif:Make=NIKON CORPORATION
- Comment: exif:MaxApertureValue=8/5
- Comment: exif:MeteringMode=3
- Comment: exif:Mode

### That is a lot more information!

* Now we have lots of information about how the photograph was created
    * Camera make and model
    * Camera settings (e.g. PhotographicSensitivity=200)
    * Note that there are values that obviously need interpretation
        * What in the world does `MeteringMode=3` mean? or `LightSource=9`?
* Notice that there is some sense of file history:
    * DateTimeOriginal=2005:07:08 19:51:47
    * DateTime=2008:12:23 12:50:21
    * modify=2017-02-06T03:47:30+00:00
    * create=2020-03-25T23:34:30+00:00
* Endianness has changed!
* There is now explicit information about how color is represented: `Pixel format: RGB`. 
    * Did TGA and TIFF only have one choice


### Now let's look at the original JPEG image

In [6]:
print(view_img_metadata("./data/daughter.jpg"))

[warn] Skip IPTC key 5: img_070905_0364


Metadata:
- Image width: 1998 pixels
- Image height: 3024 pixels
- Image orientation: Horizontal (normal)
- Bits/pixel: 24
- Pixel format: YCbCr
- Image DPI width: 72 DPI
- Image DPI height: 72 DPI
- Creation date: 2008-12-23 12:50:21
- Camera aperture: 1.6
- Camera focal: 1.8
- Camera exposure: 1/1250
- City: %G
- Camera model: NIKON D70
- Camera manufacturer: NIKON CORPORATION
- Compression: JPEG (Baseline)
- ISO speed rating: 200
- EXIF version: 0220
- Date-time original: 2005-07-08 19:51:47
- Date-time digitized: 2005-07-08 19:51:47
- Exposure bias: 0
- Focal length: 85
- Producer: QuickTime 7.5
- Comment: JPEG quality: 100%
- Format version: JFIF 1.01
- MIME type: image/jpeg
- Endianness: Big endian


### Any significant changes?

* ???

### Here's what I noticed

* Pixel format has changed from RGB to  YCbCr
* DPI matches TIFF rather than PNG

### What are the respective file sizes?

In [7]:
!ls -ltra data/daughter.*

-rw------- 1 jovyan users  2590191 Feb  6  2017 data/daughter.jpg
-rw-r--r-- 1 jovyan users 18125874 Mar 25 08:14 data/daughter.tga
-rw-r--r-- 1 jovyan users 18127819 Mar 25 08:16 data/daughter.tiff
-rw-r--r-- 1 jovyan users  4833716 Mar 26 00:05 data/daughter.png


The TIFF and TGA images are uncompressed, so they only differ by the size of the header (metadata). PNG uses a **lossless** compression, so is substantially smaller than the TIF/TGA but larger than the JPEG image which uses **lossy** compression.

### Here is a newer image

<img src="./data/skiing.jpg" alt="skiing with daughter" width="128" height="171">

In [8]:
print(view_img_metadata("./data/skiing.jpg"))

[warn] [/exif/content/ifd[0]] [Autofix] Fix parser error: stop parser, found unparsed segment: start 1432, length 8, found unparsed segment: start 1624, length 8
[warn] [/exif/content/exif[0]] [Autofix] Fix parser error: stop parser, found unparsed segment: start 12712, length 8
[warn] [/exif/content/exif_gps[0]] [Autofix] Fix parser error: stop parser, found unparsed segment: start 15800, length 8
[warn] Skip IPTC key 63: 153727
[warn] Skip IPTC key 62: 20170316


Metadata:
- Image width: 2448 pixels
- Image height: 3264 pixels
- Image orientation: Horizontal (normal)
- Bits/pixel: 24
- Pixel format: YCbCr
- Creation date: 2017-03-16 15:37:27
- Creation date: 2017-03-16 21:37:27
- Latitude: 40.58003333333333
- Longitude: -111.61911666666667
- Altitude: 2866.9 meters
- Camera focal: 2.2
- Camera exposure: 1/13333
- Camera brightness: 12.5
- City: %G
- Camera model: iPhone 6
- Camera manufacturer: Apple
- Compression: JPEG (Baseline)
- ISO speed rating: 32
- EXIF version: 0221
- Date-time original: 2017-03-16 15:37:27
- Date-time digitized: 2017-03-16 15:37:27
- Shutter speed: 13.7
- Aperture: 2.28
- Exposure bias: 0
- Focal length: 4.15
- Flashpix version: 0100
- Focal length in 35mm film: 29
- Producer: 10.2.1
- Comment: JPEG quality: 100%
- Format version: JFIF 1.01
- MIME type: image/jpeg
- Endianness: Big endian


### Notice anything new?

* ???

### What did I notice?

* Obviously difference in the camera make and model and setting
* But now we have geo location data
    - `Latitude: 40.58003333333333`
    - `Longitude: -111.61911666666667`
    - `Altitude: 2866.9 meters`
* That wasn't available in 2007! So the standards need a way of expanding/growing gracefully as technology evolves.
    * This will be important when we move on to DICOM and medical imaging standards.

## Exercise

Now let's try to reverse engineer an image standard.

You have been given 3 "images" in a proprietary format and software to render the images. You are are trying to reverse engineer the standard being used to represent the images. 

You can "peek" at the raw data in the file on the computer system and see that the files are the following list of numbers for each image respectively:

\begin{eqnarray}
1, 5, 2, 10 , 5 , 3 , 7 , -6 , 4 , 2 , 0 , 21 , 11 , -2 , 17\\
2 , 3 , 2 , 1 , 0 , 5 , -1 , 17 , 11 , -5 , 6 \\
2, 2, 3, 2, 3, 2, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144
\end{eqnarray}.

When the "images" are "rendered" to the screen, this is what you see:

\begin{equation}
\begin{array}{ccccc}
(-7,2) & (-16, -1)  & (-8,-5) & (11,1) & (-12,12)
\end{array}
\end{equation}

---------------------

\begin{equation}
\begin{array}{cc}
5 & -1 \\
17 & 11\\
-5 & 6
\end{array}
\end{equation}

------------------
\begin{equation}
\begin{array}{ccc}
(-2,-1) & (-1, 1)  & (2, 6) \\
(10, 19) & (31, 53) & (86,142)
\end{array}
\end{equation}.

------------------

The you know the image data consists of a **header**  followed by **values,** but you don't know where the **header** ends and the **values** begin. The values make up the elements (e.g. pixel--"picture element") of the picture. The header describes the structure and nature of the values. Based on your analysis, answer the following questions about the header and values


* The standard has a fixed header size (that is, the header size is the same regardless of the nature of the values) (T/F) (F)

* The values are stored by rows. T
* Image values can be multidimensional. T
* The rendered values are modified by multiplying baseline value(s) defined in the header. F. They are subtracted.



In [9]:
for q in nb1_questions:
    display(nb1_questions[q])

### [Move onto the next notebook](./dicom_intro.ipynb)